Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/polyline_snapping.h

```diff
@@ -0,0 +1,324 @@
+  Vector_3 normal = CGAL::cross_product (v0, v1);
+  if (normal == CGAL::NULL_VECTOR)
+    return false;
+  
+  Plane_3 plane0 (s0.source(), normal);
+  Plane_3 plane1 (s1.source(), normal);
+
+  Segment_3 s0proj (plane1.projection (s0.source()),
+                    plane1.projection (s0.target()));
+  Segment_3 s1proj (plane0.projection (s1.source()),
+                    plane0.projection (s1.target()));
+
+  CGAL_assertion (s0proj.source() != s0proj.target());
+  CGAL_assertion (s1proj.source() != s1proj.target());
+  
+  if (!CGAL::do_intersect (s0, s1proj) ||
+      !CGAL::do_intersect (s1, s0proj))
+    return false;
+
+  Point_3 p0, p1;
+  
+  typename CGAL::cpp11::result_of<typename Exact_kernel::Intersect_3(Segment_3, Segment_3)>::type
+    intersection0 = intersection(s0, s1proj);
+  if (!intersection0)
+    return false;
+
+  if (const Point_3* p = boost::get<Point_3>(&*intersection0))
+    p0 = *p;
+  else
+    return false;
+
+  typename CGAL::cpp11::result_of<typename Exact_kernel::Intersect_3(Segment_3, Segment_3)>::type
+    intersection1 = intersection(s1, s0proj);
+  if (!intersection1)
+    return false;
+
+  if (const Point_3* p = boost::get<Point_3>(&*intersection1))
+    p1 = *p;
+  else
+    return false;
+
+  result = CGAL::midpoint (p0, p1);
+
+  return true;
+}
+
+template <class Kernel,
+          bool Has_exact_constructions=
+          !boost::is_floating_point<typename Kernel::FT>::value >
+class Exact_snapping;
+
+template <class Kernel>
+class Exact_snapping<Kernel,false>
+{
+//typedefs
+  typedef Kernel Input_kernel;
+  typedef CGAL::Exact_predicates_exact_constructions_kernel Exact_kernel;
+  typedef CGAL::Cartesian_converter<Exact_kernel,Input_kernel> Exact_to_double;
+
+  typedef typename Input_kernel::Point_3 Inexact_point_3;
+  typedef typename Input_kernel::Line_3 Inexact_line_3;
+  typedef typename Input_kernel::Segment_3 Inexact_segment_3;
+  typedef typename Exact_kernel::Point_3 Point_3;
+  typedef typename Exact_kernel::Line_3 Line_3;
+  typedef typename Exact_kernel::Segment_3 Segment_3;
+  
+  Exact_kernel ek;
+  Exact_to_double exact_to_double;
+
+  Point_3 to_exact(const Inexact_point_3& p) const
+  {
+    return Point_3(p.x(), p.y(), p.z());
+  }
+  Inexact_point_3 to_inexact(const Point_3& p, const Exact_to_double& etd) const
+  {
+    return Inexact_point_3(etd(p.x()),
+                           etd(p.y()),
+                           etd(p.z()));
+  }
+
+public:
+
+  Exact_snapping()
+  {
+
+  }
+
+  std::pair<Inexact_point_3, bool> intersection (const Inexact_segment_3& s0, const Inexact_segment_3& s1) const
+  {
+    Segment_3 exact_s0 (to_exact (s0.source()), to_exact (s0.target()));
+    Segment_3 exact_s1 (to_exact (s1.source()), to_exact (s1.target()));
+    Point_3 result;
+    
+    if (exact_snapping<Exact_kernel> (exact_s0, exact_s1, result))
+      return std::make_pair (to_inexact (result, exact_to_double), true);
+
+    return std::make_pair (Inexact_point_3(), false);
```

Comment:
not sure it's worth is but it seems that `exact_snapping` should return an `optional<Point_3>`. I agree that you then have the conversion issue so maybe it is already the best choice.